content/en/blog/_posts/2024-08-19-pod-failure-policy-for-jobs-goes-ga.md

```diff
@@ -0,0 +1,173 @@
+
+This becomes particularly problematic when running large-scale Jobs with
+thousands of long-running Pods across thousands of nodes.
+
+The Pod failure policy extends the backoff limit mechanism to help you reduce
+costs in the following ways:
+
+- Gives you control to fail the Job as soon as a non-retriable Pod failure occurs.
+- Allows you to ignore retriable errors without increasing the `backoffLimit` field.
+
+For example, you can use a Pod failure policy to run your workload on more affordable spot machines
+by ignoring Pod failures caused by
+[graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown).
+
+The policy allows you to distinguish between retriable and non-retriable Pod
+failures based on container exit codes or Pod conditions in a failed Pod.
+
+## How it works
+
+You specify a Pod failure policy in the Job specification, represented as a list
+of rules.
+
+For each rule you define _match requirements_ based on one of the following properties:
+
+- Container exit codes: the `onExitCodes` property.
+- Pod conditions: the `onPodConditions` property.
+
+Additionally, for each rule, you specify one of the following actions to take
+when a Pod matches the rule:
+- `Ignore`: Do not count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
+- `FailJob`: Fail the entire Job and terminate all running Pods.
+- `FailIndex`: Fail the index corresponding to the failed Pod
+  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index),
+- `Count`: Count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
+  This is the default behavior.
+
+When Pod failures occur in a running Job, Kubernetes matches the
+failed Pod status against the list of Pod failure policy rules, in the specified
+order, and takes the corresponding actions for the first matched rule.
+
+### Kubernetes-initiated Pod disruptions
+
+To allow matching Pod failure policy rules against failures caused by
+disruptions initiated by Kubernetes, this feature introduces the `DisruptionTarget`
+Pod condition.
+
+This condition is added by Kubernetes when a Pod fails because of a retriable
+[disruption scenario](/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions).
+The `DisruptionTarget` condition contains one of the following reasons that corresponds to
+these disruption scenarios:
+
+- `PreemptionByKubeScheduler`: [Preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption)
+   by `kube-scheduler` to accommodate a new Pod that has a higher priority.
+- `DeletionByTaintManager` - the Pod is due to be deleted by
+   `kube-controller-manager` due to a `NoExecute` [taint](/docs/concepts/scheduling-eviction/taint-and-toleration/)
+   that the Pod doesn't tolerate.
+- `EvictionByEvictionAPI` - the Pod has been marked for
+   [eviction using the Kubernetes API](/docs/concepts/scheduling-eviction/api-eviction/).
```

Comment:
Not sure if I would say marked for eviction. Maybe something like the pod was terminated after a successful eviction.

Reply:
`DisruptionTarget` is added prior to actual eviction, and so on some rare occasions (like API server crash), the Pod may get the condition, but the actual eviction does not happen. To reflect the possibility of such scenarios we add the `Target` to the condition name, or use a tentative language like "marked for eviction". 

I admit the description may not be perfect (and some rephrasing might be helpful), but indicating as if the eviction already happened (using "was") might be misleading. wdyt?

Reply:
It all happens during a single evict request, so I am not sure there is too much difference from user's perspective.

Also if an API server crashes in the middle of the processing, then there is no guarantee the eviction will continue. It needs to be initiated again by another actor. That is why `marked for eviction` seems a bit confusing to me.

Reply:
Yes, it happens within a single request, but what I was told by APi folks, IIUC there is no guarantee on atomicity of adding the condition, and actual deletion. cc @liggitt to keep me honest here.

So, in that case the actual eviction may not actually happen. For this reason we introduce a mechanism to clean stale DisruptionTarget conditions. For this reason we introduce the cleaner of stale conditions which removes them after 2min (see [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#stale-disruptiontarget-condition-which-is-not-cleaned-up)). Unfortunately this is not in user-facing docs currently, and I don't think it will make into this release cycle. I can add a note in the blog post about it. OTOH this is very much going into the weeds / details. WDYT?

Reply:
I did not mean to dive into the logic here or in the blog. I was just wondering if we can come up with a better sentence, but I can live with the current version as well :)

Reply:
what about `the Pod is to be terminated due to an API-initiated eviction`?

Reply:
I see, still for curiosity, and to refresh my memory, I did a bit of code search on my own and indeed it seems like two calls `Update` and `Delete` (see [here](https://github.com/kubernetes/kubernetes/blob/dbc2b0a5c7acc349ea71a14e49913661eaf708d2/pkg/registry/core/pod/storage/eviction.go#L307-L362)), so I believe there is no atomicity.

As for the wording, the suggested is ok, but I updated to "the Pod is due to be deleted by an [API-initiated eviction]" to also align with the wording for other scenarios. PTAL.